src/Microsoft.AspNet.CookiePolicy/CookiePolicyMiddleware.cs

```diff
@@ -0,0 +1,157 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+using Microsoft.AspNet.Builder;
+using Microsoft.AspNet.Http;
+using Microsoft.AspNet.Http.Features;
+using Microsoft.AspNet.Http.Features.Internal;
+
+namespace Microsoft.AspNet.CookiePolicy
+{
+    public class CookiePolicyMiddleware
+    {
+        private readonly RequestDelegate _next;
+
+        public CookiePolicyMiddleware(
+            RequestDelegate next,
+            CookiePolicyOptions options)
+        {
+            Options = options;
+            _next = next;
+        }
+
+        public CookiePolicyOptions Options { get; set; }
+
+        public Task Invoke(HttpContext context)
+        {
+            // REVIEW: Do we need to check if there is a Cookie feature already present like SendFile??
+            context.Features.Set<IResponseCookiesFeature>(new CookiesWrapperFeature(context, Options));
+
+            return _next(context);
+        }
+
+        private class CookiesWrapperFeature : IResponseCookiesFeature
+        {
+            public CookiesWrapperFeature(HttpContext context, CookiePolicyOptions options)
+            {
+                Wrapper = new CookiesWrapper(context, options, context.Response.Cookies);
+            }
+
+            public IResponseCookies Wrapper { get; }
+
+            public IResponseCookies Cookies
+            {
+                get
+                {
+                    return Wrapper;
+                }
+            }
+        }
+
+        private class CookiesWrapper : IResponseCookies
+        {
+            public CookiesWrapper(HttpContext context, CookiePolicyOptions options, IResponseCookies cookies)
+            {
+                Context = context;
+                Cookies = cookies;
+                Policy = options;
+            }
+
+            public HttpContext Context { get; }
+
+            public IResponseCookies Cookies { get; }
+
+            public CookiePolicyOptions Policy { get; }
+
+            private bool PolicyRequiresCookieOptions()
+            {
+                return Policy.HttpOnly != HttpOnlyPolicy.None || Policy.Secure != SecurePolicy.None;
+            }
+
+            public void Append(string key, string value)
+            {
+                if (PolicyRequiresCookieOptions())
+                {
+                    Append(key, value, new CookieOptions());
+                    return;
+                }
+
+                if (Policy.OnAppendCookie != null)
+                {
+                    var context = new AppendCookieContext(Context, options: null, name: key, value: value);
+                    Policy.OnAppendCookie(context);
+                    key = context.CookieName;
+                    value = context.CookieValue;
+                }
+                Cookies.Append(key, value);
+            }
+
+            public void Append(string key, string value, CookieOptions options)
```

Comment:
There didn't appear to be any argument validation in the base HttpAbstractions ResponseCookies so I just mimiced that behavior since we are just wrapping no?